Handle container registry authentication via userdata

bpradipt opened the pull request:
This patch series includes the following:
1. Adds support for providing the registry authentication file via userdata. 
2. Removes usage of multiple files via cloud-config. All required values are part of daemon Config which is then processed by process-user-data command inside podvm
3. Adds support for fetching of userdata from metadata service in AWS.
4. Adds some misc improvements and tests to the AWS code

snir911 commented on pkg/adaptor/cloud/aws/provider.go:
```diff
@@ -226,6 +226,21 @@ func (p *awsProvider) CreateInstance(ctx context.Context, podName, sandboxID str
+				AmdSevSnp: types.AmdSevSnpSpecificationEnabled,
+			}
+		}
+
```
should default instance type also adapt to it?

bpradipt replied:
Good point. I'll add it.

snir911 commented on pkg/adaptor/cloud/aws/types.go:
```diff
@@ -17,7 +17,11 @@ func (i *securityGroupIds) String() string {
 func (i *securityGroupIds) Set(value string) error {
-	*i = append(*i, strings.Split(value, ",")...)
+	if len(value) == 0 {
+		*i = make(securityGroupIds, 0)
```
is it needed? AFAIR append works on nil slice

bpradipt replied:
Hmm.. I'll recheck

bpradipt commented on entrypoint.sh:
```diff
@@ -47,6 +47,7 @@ aws() {
     [[ "${TAGS}" ]] && optionals+="-tags ${TAGS} " # Custom tags applied to pod vm
     [[ "${USE_PUBLIC_IP}" == "true" ]] && optionals+="-use-public-ip " # Use public IP for pod vm
     [[ "${ROOT_VOLUME_SIZE}" ]] && optionals+="-root-volume-size ${ROOT_VOLUME_SIZE} " # Specify root volume size for pod vm
+    [[ "${DISABLE_CLOUD_CONFIG}" == "true" ]] && optionals+="-disable-cloud-config "
```
Currently disabling of cloud config is only for aws and azure. And since the option in the global sections are generic, moving cloud config configuration to global will be confusing imho. 
May be once we figure out how to implement it for other providers, especially ibmcloud I'll make this part of cloud config and move it under the common section. Does that sound ok ?

snir911 reviewed (approved):
LGTM thanks!

stevenhorsman commented on cmd/process-user-data/types.go:
```diff
@@ -10,6 +10,10 @@ const (
 
 	defaultAgentConfigPath = "/etc/agent-config.toml"
+
+	authJsonDirPath    = "/etc/attestation-agent/"
```
nit: The auth.json is an image-rs config now, so I'm not sure if `/etc/attestation-agent/` is the best directory rather than just `/etc/`, but it's not a blocker

bpradipt replied:
I can change the directory. Thought `attestation-agent` might be a good place to keep all these auth stuff. I'll change to `/etc` to align with the agent-config.toml location. In future we can decide to put all these configs under `/etc/kata` may be.

stevenhorsman replied:
Yeah, I wonder if, as part of the merge into main, we might end up adopting policies to set some of this config rather than the current approach anyway.

stevenhorsman commented on cmd/process-user-data/update.go:
```diff
@@ -106,6 +107,39 @@ func updateAgentConfig(cmd *cobra.Command, args []string) error {
+				return fmt.Errorf("failed to write agent config file: %s", err)
+			}
+
+		} else if strings.Contains(agentConfig.AaKbcParams, "cc_kbc") {
```
I'm wondering if we should just use the `image_registry_auth_file=file:///etc/auth.json` config option for all kbcs so we have less branching and more shared code as it should still be valid?

bpradipt replied:
Does it apply for all KBCs ? I'm open for changing

stevenhorsman replied:
So my understanding (which could be wrong), is that by specifying `image_registry_auth_file=file:///etc/auth.json` image-rs uses that file directly, rather than going to the attestation-agent (and therefore KBC) to get that information. I'm happy to test it out trying to build a peer-pod with offline_fs_kbc if you make the code general

stevenhorsman reviewed (commented):
A lot of this PR is AWS related, so I don't have too much to contribute there, but I have a couple of questions on the shared piece and after we've decided on them I can give it a test on ibm cloud (depending on the result)

mkulke commented on cmd/process-user-data/update.go:
```diff
@@ -94,14 +96,92 @@ func updateAgentConfig(cmd *cobra.Command, args []string) error {
+			} else {
+				fmt.Printf("auth.json file doesn't exist. Not updating the image_registry_auth_file in agent config file\n")
+			}
+		} else {
+			fmt.Printf("Unknown KBC option. Not updating the agent config file\n")
+		}
```
this is quite nested. can we factor that into it's own function, .e.g `setAuthJsonConfig(agentConfig, config.AuthJson)`?

bpradipt replied:
Yeah, @stevenhorsman also suggested to set this for any kbc.

mkulke commented on cmd/process-user-data/update.go:
```diff
@@ -94,14 +96,92 @@ func updateAgentConfig(cmd *cobra.Command, args []string) error {
+	err = toml.Unmarshal(data, &agentConfig)
+	if err != nil {
+		fmt.Println("Error parsing agent config file:", err)
+		return AgentConfig{}, err
```
why return an empty AgentConfig instead of `nil`? would limit the risk of someone misusing it on the caller site

bpradipt replied:
ok, I'll change the method

bpradipt commented on cmd/process-user-data/provision.go:
```diff
@@ -45,21 +45,53 @@ func isAzure(ctx context.Context) bool {
 	}
 
-	return userDataUrl
+	if isAWS(ctx) {
```
We'll need heuristics if using generic qcow2/raw images. But if we are building native images, then this heuristic is not needed and probably set via ldflags. Should I explore this approach as well in this PR or a next one ?

mkulke replied:
I see. We maybe want to employ some other heuristics to avoid the http roundtrips:

there are a set of [methods for AWS](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html)

for azure the prescribes dmidecode method via smbios should also work: `sudo /usr/sbin/dmidecode -s {system-product-name,system-manufactorer}`. We could possibly retrieve this by using a go library: https://blog.gopheracademy.com/advent-2017/accessing-smbios-information-with-go/

bpradipt replied:
The dmidecode parser code has not been updated since many years. Other option is to directly parse `/sys` to get some of these info. But reading the document, this is also not fool proof.
As for AWS, the definitive way is to get the `instance identity document` from the reference doc you mentioned. I'm using that to check if it's AWS. Likewise for Azure, using what is described in https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=linux

@mkulke I'm thinking of continuing with additional exploration to use `/sys` or compile time option for provider specific builds in a future PR. Hope this is fine with you

mkulke replied:
Yes, let's open an issue as a reminder about the concern with the current approach, though.

mkulke replied:
Yes, let's open an issue as a reminder about the concern with the current approach, though.

bpradipt commented on cmd/process-user-data/provision.go:
```diff
@@ -145,6 +178,52 @@ func getUserData(ctx context.Context, url string) (string, error) {
 
+// Add a method to retrieve userData from the instance metadata service
+// and return it as a string
+func getUserDataForAws(ctx context.Context, url string) (string, error) {
```
yeah

bpradipt commented on cmd/process-user-data/provision.go:
```diff
@@ -212,14 +291,22 @@ func provisionFiles(cmd *cobra.Command, args []string) error {
 
 	err := retry.Do(
 		func() error {
-			cfg.userData, _ = getUserData(ctx, userDataUrl)
```
ok

bpradipt commented:
@stevenhorsman @mkulke I have addressed all the comments, except one on detecting AWS or Azure for fetching the user data - https://github.com/confidential-containers/cloud-api-adaptor/pull/1440#discussion_r1331966005
PTAL

bpradipt commented:
@stevenhorsman I have tested with both offline_fs_kbc and cc_kbc in AWS. If you can test it in ibmcloud, it'll be great

stevenhorsman commented:
> @stevenhorsman I have tested with both offline_fs_kbc and cc_kbc in AWS. If you can test it in ibmcloud, it'll be great

Sure - I'm just focussed on the secrets fix now, but will try and get to this later today and please remind me if I don't

stevenhorsman commented:
I build a new podvm based on this code and (assuming that I did it correctly whilst juggling the secrets stuff), is seems to be working well for ibmcloud with offline_fs_kbc on s390x:
```
=== RUN   TestCreatePeerPodWithAuthenticatedImagewithValidCredentials
=== RUN   TestCreatePeerPodWithAuthenticatedImagewithValidCredentials/ValidAuthImagePeerPod_test
=== RUN   TestCreatePeerPodWithAuthenticatedImagewithValidCredentials/ValidAuthImagePeerPod_test/Peer_pod_with_Authenticated_Image_with_Valid_Credentials_has_been_created
    common_suite_test.go:266: PodVM has successfully reached Completed state with authenticated Image - quay.io/kata-containers/confidential-containers-auth:test
time="2023-09-21T09:19:04-07:00" level=info msg="Deleting Secret... auth-json-secret"
time="2023-09-21T09:19:04-07:00" level=info msg="Deleting pod authenticated-image-valid-3530584652-pod..."
--- PASS: TestCreatePeerPodWithAuthenticatedImagewithValidCredentials (28.99s)
    --- PASS: TestCreatePeerPodWithAuthenticatedImagewithValidCredentials/ValidAuthImagePeerPod_test (28.99s)
        --- PASS: TestCreatePeerPodWithAuthenticatedImagewithValidCredentials/ValidAuthImagePeerPod_test/Peer_pod_with_Authenticated_Image_with_Valid_Credentials_has_been_created (22.58s)
=== RUN   TestCreatePeerPodWithAuthenticatedImageWithInvalidCredentials
=== RUN   TestCreatePeerPodWithAuthenticatedImageWithInvalidCredentials/InvalidAuthImagePeerPod_test
=== RUN   TestCreatePeerPodWithAuthenticatedImageWithInvalidCredentials/InvalidAuthImagePeerPod_test/Peer_pod_with_Authenticated_Image_with_Invalid_Credentials_has_been_created
    common_suite_test.go:266: PodVM has successfully reached ImagePullBackOff state with authenticated Image - quay.io/kata-containers/confidential-containers-auth:test
time="2023-09-21T09:19:36-07:00" level=info msg="Deleting Secret... auth-json-secret"
time="2023-09-21T09:19:36-07:00" level=info msg="Deleting pod authenticated-image-invalid-1951475890-pod..."
--- PASS: TestCreatePeerPodWithAuthenticatedImageWithInvalidCredentials (32.39s)
    --- PASS: TestCreatePeerPodWithAuthenticatedImageWithInvalidCredentials/InvalidAuthImagePeerPod_test (32.39s)
        --- PASS: TestCreatePeerPodWithAuthenticatedImageWithInvalidCredentials/InvalidAuthImagePeerPod_test/Peer_pod_with_Authenticated_Image_with_Invalid_Credentials_has_been_created (26.00s)
=== RUN   TestCreatePeerPodWithAuthenticatedImageWithoutCredentials
=== RUN   TestCreatePeerPodWithAuthenticatedImageWithoutCredentials/InvalidAuthImagePeerPod_test
time="2023-09-21T09:19:37-07:00" level=info msg="Deleting pre-existing auth-json-secret..."
time="2023-09-21T09:19:37-07:00" level=info msg="Creating empty auth-json-secret..."
=== RUN   TestCreatePeerPodWithAuthenticatedImageWithoutCredentials/InvalidAuthImagePeerPod_test/Peer_pod_with_Authenticated_Image_with_Invalid_Credentials_has_been_created
    common_suite_test.go:266: PodVM has successfully reached WithoutCredentials state with authenticated Image - quay.io/kata-containers/confidential-containers-auth:test
time="2023-09-21T09:20:06-07:00" level=info msg="Deleting pod authenticated-image-without-creds-1263014715-pod..."
--- PASS: TestCreatePeerPodWithAuthenticatedImageWithoutCredentials (29.88s)
    --- PASS: TestCreatePeerPodWithAuthenticatedImageWithoutCredentials/InvalidAuthImagePeerPod_test (29.88s)
        --- PASS: TestCreatePeerPodWithAuthenticatedImageWithoutCredentials/InvalidAuthImagePeerPod_test/Peer_pod_with_Authenticated_Image_with_Invalid_Credentials_has_been_created (23.19s)
PASS
```
